### PR TITLE
Migrating the AG news and BoolQ Notebooks

### PR DESCRIPTION
**Pull request by emersodb**
Migrating the AG News and Bool Q notebooks. Editing Config readme, fixing a bug with metrics calculation, cleaning AG news dataset and BoolQ dataset examples. Removing superfluous init.

Clickup Tickets: 

[Ticket 1](https://app.clickup.com/t/8686z40g7), [Ticket 2](https://app.clickup.com/t/8686z458z)

### CODE REVIEW

File: src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb [hunk 1]
```diff
@@ -0,0 +1,1107 @@
...
+    "    print(text)\n",
+    "    print(\"==================================\")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Few-shot learning definitely helps a lot! We'll measure accuracy on a sample of the AG news dataset below. However, there is nothing stoping the model from not selecting our labels. So can we do better? We can work around this by understanding the likelihood of our labels from the models perspective. This will also allow us to use zero-shot learning, even when the model doesn't seem to want to respond in the way we expect."
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb]: "stoping" -> "stopping", "models" -> "model's"

File: src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb [hunk 2]
```diff
@@ -0,0 +1,1107 @@
...
+    "last_layer_name = model.module_names[-1]\n",
+    "last_layer_name"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The last layer of the model is analogous to the probabilities of each token in the model vocabulary. That is, it is the conditional probability\n",
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb]: I think we can rewrite this paragraph a bit better. For example: 

"last layer of the model" -> "last layer activations of the model"

"we get back a 32K vector..." -> "we get back a vector of dimension 32000 (the vocabulary size of Llama-2) which represents the probability distribution of  $y_{t+1}$ over the vocabulary..."

File: src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb [hunk 3]
```diff
@@ -0,0 +1,1107 @@
...
+    "print(f\"Number of tokens: {len(tokenizer.encode(single_prompted_input))}\")\n",
+    "# The shape of this tensor should be number of input tokens by the vocabulary size (n x 32000)\n",
+    "print(f\"Activations matrix shape: {last_layer_matrix.shape}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We're interested in the largest logits (pre-softmax projections) from our label collection. The function `get_label_with_highest_likelihood` looks into the last value in the activations matrix (corresponding to the probability of the first predicted token) and finds the largest logits among our labels."
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb]: I think we can rewrite this paragraph to be a bit clearer too. Something like: 

"We're interested in the logits (i.e., activations prior to applying softmax) which correspond to our labels. The function `get_label_with_highest_likelihood` looks into the last row of the activations matrix (corresponding to the probability distribution over the vocabulary of the first predicted token) and finds the largest logit among our labels."

File: src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb [hunk 4]
```diff
@@ -0,0 +1,1107 @@
...
+    "print(f\"Failed to match {len(unmatched_predictions)} responses to our label space\")\n",
+    "print(f\"Some examples of responses: {sample(unmatched_predictions, 10)}\")\n",
+    "report_metrics(predicted_labels, ag_news_labels, labels_order=[\"world\", \"sports\", \"business\", \"technology\"])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can see that the model actually never answers with a response in our expected label space. As a result, we end up randomly guessing for all of our predictions, leading to random accuracy."
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb]: I feel like "random accuracy" might be a bit misleading 😂

File: src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb [hunk 5]
```diff
@@ -0,0 +1,1107 @@
...
+   ],
+   "source": [
+    "report_metrics(predicted_labels, ag_news_labels, labels_order=[\"world\", \"sports\", \"business\", \"technology\"])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "There are a few examples where the model doesn't answer in the space we expect, but there are very few. The accuracy is a significant improvement over zero-shot."
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb]: Maybe we can reword "but there are very few" to "but there aren't many such cases"

File: src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb [hunk 6]
```diff
@@ -0,0 +1,1107 @@
...
+   ],
+   "source": [
+    "report_metrics(predicted_labels, ag_news_labels, labels_order=[\"world\", \"sports\", \"business\", \"technology\"])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This is a large jump in the accuracy of our predictions for zero-shot. Because we're directly extracting probabilities associated with our label and selecting the highest, we have no issues with matching. Moreover, we're actually able to match the performance of few-shot prompts with just zero-shot. It should be noted that part of this is likely the fact that we're using `temperature=1.0` for the few-shot prompting, but it's still a big improvement. "
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb]: Suggestions for rephrasing this:

"We can see this approach yielded a large improvement in prediction accuracy over the naive zero-shot prompting approach. Because we are only extracting the probabilities that match our label space, we do not have the issue with the prediction being outside of the label space."

"part of this is likely the fact that" -> "this is likely in part due to the fact that"

Comment (yuchongzhang) [reply src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb]: I think it can be helpful to say a bit more why we might expect `temperature=1.0` to be a cause

File: src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb [hunk 7]
```diff
@@ -0,0 +1,1107 @@
...
+   ],
+   "source": [
+    "report_metrics(predicted_labels, ag_news_labels, labels_order=[\"world\", \"sports\", \"business\", \"technology\"])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Because we're using the likelihood mapping, there are no instances in which we fail to match our labels and we're performing this task quite well. Note that if we set our generation to greedy decoding (`temperature = 0`), we would likely get close to this performance without likelihood matching, but we'd still have label matching issues for some of our generations."
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb:1081]: Similarly, we can say a bit more why we expect greedy decoding to have this effect

File: src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_boolq.ipynb [hunk 8]
```diff
@@ -176,44 +220,90 @@
...
    "metadata": {},
-   "outputs": [],
    "source": [
-    "# We're interested in the activations from the last layer of the model, because this will allow us to calculate the\n",
-    "# likelihoods.\n",
+    "We're also going to use model activations in this task. The last layer of the model is analogous to the probabilities of each token in the model vocabulary. That is, the conditional probability\n",
+    "$$\n",
+    "P(y_t \\vert y_{<t}, x),\n",
+    "$$\n",
+    "The probability distribution over the vocabulary of the next token given the preceding tokens $y_{<t}$, and the prompt text $x$. Thus, for each token $y_{t}$ in our input, we get back a 32K vector corresponding to the probabilities over the vocabulary of $y_{t+1}$. For this example, we only care about the last token in our input, as it houses the probability of the, as yet, unseen token the model will generate.\n",
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_boolq.ipynb]: See [here](https://github.com/VectorInstitute/PromptEngineeringLaboratory/pull/1/files#r1514895686)

File: src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_boolq.ipynb [hunk 9]
```diff
@@ -222,279 +312,677 @@
...
+    }
+   ],
+   "source": [
+    "label_map = {0: \"False\", 1: \"True\"}\n",
     "label_ordering = [\"False\", \"True\"]\n",
-    "demonstrations_1 = create_demonstrations(\n",
-    "    bool_q_train_titles, bool_q_train_passages, bool_q_train_questions, bool_q_train_labels, label_map_1, 1\n",
-    ")\n",
-    "prompts_1 = create_prompts(demonstrations_1, bool_q_test_titles, bool_q_test_passages, bool_q_test_questions)"
+    "lowercase_labels = [label.lower() for label in label_ordering]\n",
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_boolq.ipynb:677]: Super minor: we are not consistent in whether to use lower case labels or not. Just wanted to leave a comment on this in case you think it's a good idea to make changes about it.

**Comment by yuchongzhang** [comment]
Most of my comments are about rephrasing certain sentences or adding a bit more details. Feel free to ignore if you think what is already there is sufficient. Everything else looks good to me!

File: src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb (hunk 4 above)
Comment (emersodb) [reply src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb]: That's fair 😂

File: src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb (hunk 6 above)
Comment (emersodb) [reply src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb]: Done!

File: src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_boolq.ipynb (hunk 9 above)
Comment (emersodb) [reply src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_boolq.ipynb:677]: Yeah I know what you mean. I'm careful about it to make sure it doesn't impact performance, but I can see how it might be confusing. However, in the name of time, I'm going to leave it for now and put a note to myself to address it later if there is time.

**Review by yuchongzhang — APPROVED** [review]
